collections/collections.go

```diff
@@ -0,0 +1,104 @@
+// meaning that having two namespaces whose bytes representation is:
+// p1 := []byte("prefix")
+// p2 := []byte("prefix1")
+// yields to iterations of p1 overlapping over p2.
+// If a numeric prefix is provided, it must be between 0 and 255 (uint8).
+// If out of bounds this function will panic.
+// Reason for which this function is constrained to `int` instead of `uint8` is for
+// API ergonomics, golang's type inference will infer int properly but not uint8
+// meaning that developers would need to write NewPrefix(uint8(number)) for numeric
+// prefixes.
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		identifierCopy := make([]byte, len(c))
+		copy(identifierCopy, c)
+		prefix = identifierCopy
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyCodec defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyCodec[T any] interface {
+	// Encode writes the key bytes into the buffer. Returns the number of
+	// bytes written. The implementer must expect the buffer to be at least
+	// of length equal to Size(K) for all encodings.
+	// It must also return the number of written bytes which must be
+	// equal to Size(K) for all encodings not involving varints.
+	// In case of encodings involving varints then the returned
+	// number of written bytes is allowed to be smaller than Size(K).
+	Encode(buffer []byte, key T) (int, error)
+	// Decode reads from the provided bytes buffer to decode
+	// the key T. Returns the number of bytes read, the type T
+	// or an error in case of decoding failure.
+	Decode(buffer []byte) (int, T, error)
+	// Size returns the buffer size need to encode key T in binary format.
+	// The returned value must match what is computed by EncodeKey for all
+	// encodings except the ones involving varints. Varints are expected
+	// to return the maximum varint bytes buffer length, at the risk of
+	// over-estimating in order to pick the most performant path.
+	Size(key T) int
+	// Stringify returns a string representation of T.
+	Stringify(key T) string
+	// KeyType returns a string identifier for the type of the key.
+	KeyType() string
+}
+
+// ValueCodec defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collection values.
+type ValueCodec[T any] interface {
+	// Encode encodes the value T into binary format.
+	Encode(value T) ([]byte, error)
+	// Decode returns the type T given its binary representation.
+	Decode(b []byte) (T, error)
+	// Stringify returns a string representation of T.
+	Stringify(value T) string
+	// ValueType returns the identifier for the type.
+	ValueType() string
+}
```

Comment:
KeyCodec defines an Encode method that writes to a buffer provided by callers. ValueCodec defines an EncodeMethod that _returns_ a slice of bytes. Why are they different?

Reply:
@peterbourgon the reality is that the KVStore bytes key we use to map an object is composed, it contains:
- a bytes prefix owned by collections.Map which identifies the object's storage namespace.
- a key returned by `KeyEncoder.Encode(K)`

so the KVStore key to map an object is `valueSavingKey := append(m.prefix, KeyEncoder.Encode(K))`
in order to make this performant, we pre-allocate the whole key buffer

On top of this, a `KeyEncoder` can be a composition of multiple `KeyEncoder` instances, eg:

```go
type Pair[K1, K2] struct { k1 *K1, k2 *K2 }
type PairKeyEncoder[K1, K2] struct {
   keyEncoder1 KeyEncoder[K1]
   keyEncoder2 KeyEncoder[K2]
}
```

The API can then properly understand, during decoding, how many bytes is K1 and how many bytes is K2.

Value is fundamentally different, it does not have the keys properties and it's self contained in its bytes provision.

Reply:
> If you want to talk about performance, I'm down! Let's start at a profile of a typical Cosmos app and drill down.

Happy to engage in this anytime :)

Reply:
I think we are probably talking past each other. I think you're trying to say that collections.Map captures a KVStore + some prefix + some encoding of key and value types. I get that.